textproto/header.go

```diff
@@ -33,19 +33,35 @@ type Header struct {
 	// Fields are in reverse order so that inserting a new field at the top is
 	// cheap.
 	l []*headerField
-	m map[string][]*headerField
+
+	// Contains first field value we seen.
+	// Actually, first value can be stored in mDups instead of this map. See
+	// headerFields.Del for details.
+	m map[string]*headerField
+	// Second and further values.
+	mDups map[string][]*headerField
```

Comment:
Hmm. I'm not sure how I feel about this. `Header` is already quite complicated, maybe we could extract the logic into `MultiStringMap` or something? Maybe we could simplify the patch by either using `m` if there's a single value, either using `mDups` if there are multiple but avoid having to split values across both?

Maybe we can merge the other commits, and discuss this in a separate PR?

Reply:
>Maybe we could simplify the patch by either using m if there's a single value, either using mDups if there are multiple but avoid having to split values across both?

I guess that's probably the better way. I will take a look. For now, I removed that commit from this PR.